### ISSUE TITLE
loading cvt individuals

### ISSUE THREAD

**Issue by bossdm** [resilient-swarms/argos-sferes#9]
When I try loading an individual with cvt-based descriptor, it seems to fail. The output is the same as if the individual is not being found: in this case the generation file is loaded correctly, but the run stops after loading.

For example,

` ./bin/obsavoid_evolcvt42D experiments/mutualinfo.argos --load experiments/gen_1 -o outputfile -n 37
`
results in the output:
`[INFO] Not using threads
[INFO] Using random seed = 56
[INFO] Using simulation clock tick = 0.1
[INFO] Total experiment length in clock ticks = 1200
[INFO] Loaded library "./lib/libobsavoid_nn_controller.so"
[INFO] Loaded library "./lib/libobsavoid_evol_loopfunctionscvt42D.so"
Range = 0.11[INFO] The physics engine "dyn2d" will perform 10 iterations per tick (dt = 0.01 sec)
[INFO] No visualization selected.

Loaded 1000 centroids.
sferes2 version: (const char*)"f731bfb04a48a475dfedbfdb180d77054498557f"
seed: 1557888548
loading experiments/gen_1
`

**Comment by daneshtarapore** [comment]
Can you send me the population files so that I can try and replicate your error.

**Comment by daneshtarapore** [comment]
Am having difficulty replicating your error:

danesh@danesh-UOS214035:~/argos-sferes$ ./bin/obsavoid_evolcvt42D experiments/mutualinfo.argos --load experiments/gen_100 -o outputfile -n 14
[INFO] Not using threads
[INFO] Using random seed = 56
[INFO] Using simulation clock tick = 0.1
[INFO] Total experiment length in clock ticks = 1200
[INFO] Loaded library "./lib/libobsavoid_nn_controller.so"
[INFO] Loaded library "./lib/libobsavoid_evol_loopfunctionscvt42D.so"
Position1 2.56518,3.75618,0 trial 0 time 0
Position1 1.07234,1.90009,0 trial 1 time 0
Position1 3.19485,3.9278,0 trial 2 time 0
Position1 3.02592,1.1045,0 trial 3 time 0
Position1 2.95131,3.04298,0 trial 4 time 0
position after reset 2.56518,3.75618,0[INFO] The physics engine "dyn2d" will perform 10 iterations per tick (dt = 0.01 sec)
[INFO] No visualization selected.
experiments/centroids_1000_42.dat
Error: Could not load the centroids.

Any idea where this error is coming from?

**Comment by bossdm** [comment]
maybe do a git pull again now, the example centroid initialisations are now included in the experiments folder. however, something to think about is to remove the necessity to load this file when resuming/testing from a generation file.

**Comment by bossdm** [comment]
Not sure if it is related but configuring map_elites and cvt_map_elites
gives some problems:
`./waf configure --no-mpi`
gives the following output:
`WARNING simplejson not found some function may not work
Command-line options for module [modules/nn2] :  -> no option found 
Command-line options for module [modules/cvt_map_elites] : OK 
Command-line options for module [modules/map_elites]     :  -> no option found 
Command-line options for exp [exp/example]               :  -> OK 
Setting top to                           : /home/david/argos-sferes/sferes2 
Setting out to                           : /home/david/argos-sferes/sferes2/build 
Checking for 'g++' (C++ compiler)        : /usr/bin/g++ 
Checking boost includes                  : 1_70 
Checking boost libs                      : ok 
Checking Intel TBB includes (optional)   : /usr/include 
Checking Intel TBB libs (optional)       : /usr/lib/x86_64-linux-gnu 
Checking for Eigen                       : ok 
Checking pthread                         : /usr/lib/x86_64-linux-gnu 
Configuring for module: [modules/nn2]
modules/nn2 -> ok
Configuring for module: [modules/cvt_map_elites]
modules/cvt_map_elites -> ok
Configuring for module: [modules/map_elites]
--- Logging error ---
Traceback (most recent call last):
  File "/home/david/argos-sferes/sferes2/wscript", line 168, in configure
    conf.recurse(i)
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Context.py", line 125, in recurse
    raise Errors.WafError('No function %r defined in %s'%(name or self.fun,node.abspath()))
waflib.Errors.WafError: No function 'configure' defined in /home/david/argos-sferes/sferes2/modules/map_elites/wscript

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 1034, in emit
    msg = self.format(record)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 880, in format
    return fmt.format(record)
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Logs.py", line 134, in format
    return msg%rec.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "./waf", line 165, in <module>
    Scripting.waf_entry_point(cwd, VERSION, wafdir)
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Scripting.py", line 119, in waf_entry_point
    run_commands()
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Scripting.py", line 179, in run_commands
    ctx=run_command(cmd_name)
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Scripting.py", line 170, in run_command
    ctx.execute()
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Configure.py", line 85, in execute
    super(ConfigurationContext,self).execute()
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Context.py", line 85, in execute
    self.recurse([os.path.dirname(g_module.root_path)])
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Context.py", line 126, in recurse
    user_function(self)
  File "/home/david/argos-sferes/sferes2/wscript", line 171, in configure
    Logs.warn(' %s -> no configuration found' % i, 'YELLOW')
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Logs.py", line 160, in warn
    log.warn(*k,**kw)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 1400, in warn
    self.warning(msg, *args, **kwargs)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 1395, in warning
    self._log(WARNING, msg, args, **kwargs)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 1519, in _log
    self.handle(record)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 1529, in handle
    self.callHandlers(record)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 1591, in callHandlers
    hdlr.handle(record)
  File "/home/david/anaconda3/lib/python3.7/logging/__init__.py", line 905, in handle
    self.emit(record)
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Logs.py", line 84, in emit
    self.emit_override(record)
  File "/home/david/argos-sferes/sferes2/.waf3-2.0.14-e67604cd8962dbdaf7c93e0d7470ef5b/waflib/Logs.py", line 108, in emit_override
    logging.StreamHandler.emit(self,record)
Message: ' modules/map_elites -> no configuration found'
Arguments: ('YELLOW',)

--- configuration ---
compiler(s):
 * CXX: gcc
boost version: 1_70
mpi: False
Compilation flags :
   CXXFLAGS : -D_REENTRANT -Wall -fPIC -ftemplate-depth-1024 -Wno-sign-compare -Wno-deprecated  -Wno-unused -DSFERES_ROOT="/home/david/argos-sferes/sferes2"  -std=c++11 -DEIGEN3_ENABLED 
   LINKFLAGS: 
--- license ---
Sferes2 is distributed under the CECILL license (GPL-compatible)
Please check the accompanying COPYING file or http://www.cecill.info/
'configure' finished successfully (0.138s)
`

and `./waf build -v` gives the following output:


`WARNING simplejson not found some function may not work
Command-line options for module [modules/nn2] :  -> no option found 
Command-line options for module [modules/cvt_map_elites] : OK 
Command-line options for module [modules/map_elites]     :  -> no option found 
Command-line options for exp [exp/example]               :  -> OK 
Waf: Entering directory `/home/david/argos-sferes/sferes2/build'
DEBUG is is disabled
Entering directory `/home/david/argos-sferes/sferes2'
Building module: modules/nn2
Entering directory `/home/david/argos-sferes/sferes2/modules/'
Building module: modules/cvt_map_elites
Building module: modules/map_elites
Entering directory `/home/david/argos-sferes/sferes2/modules/'
[ 5/34] Linking build/modules/map_elites/test_map_elites
17:11:08 runner ['/usr/bin/g++', 'modules/map_elites/test_map_elites.cpp.1.o', '-o/home/david/argos-sferes/sferes2/build/modules/map_elites/test_map_elites', '-Wl,-Bstatic', '-Lsferes', '-lsferes2', '-Wl,-Bdynamic', '-L/usr/lib', '-L/usr/lib/x86_64-linux-gnu', '-lboost_serialization', '-lboost_filesystem', '-lboost_system', '-lboost_unit_test_framework', '-lboost_program_options', '-lboost_graph', '-lboost_thread', '-lboost_regex', '-ltbb']
[31/34] Linking build/modules/cvt_map_elites/test_cvt_map_elites
17:11:08 runner ['/usr/bin/g++', 'modules/cvt_map_elites/test_cvt_map_elites.cpp.1.o', '-o/home/david/argos-sferes/sferes2/build/modules/cvt_map_elites/test_cvt_map_elites', '-Wl,-Bstatic', '-Lsferes', '-lsferes2', '-Wl,-Bdynamic', '-L/usr/lib', '-L/usr/lib/x86_64-linux-gnu', '-lboost_serialization', '-lboost_filesystem', '-lboost_system', '-lboost_unit_test_framework', '-lboost_program_options', '-lboost_graph', '-lboost_thread', '-lboost_regex', '-ltbb']
/usr/bin/ld: modules/map_elites/test_map_elites.cpp.1.o: undefined reference to symbol 'pthread_join@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status

/usr/bin/ld: modules/cvt_map_elites/test_cvt_map_elites.cpp.1.o: undefined reference to symbol 'pthread_join@@GLIBC_2.2.5'
//lib/x86_64-linux-gnu/libpthread.so.0: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status

Waf: Leaving directory `/home/david/argos-sferes/sferes2/build'
Build failed
 -> task in 'test_map_elites' failed with exit status 1: 
	{task 140195946492536: cxxprogram test_map_elites.cpp.1.o -> test_map_elites}
['/usr/bin/g++', 'modules/map_elites/test_map_elites.cpp.1.o', '-o/home/david/argos-sferes/sferes2/build/modules/map_elites/test_map_elites', '-Wl,-Bstatic', '-Lsferes', '-lsferes2', '-Wl,-Bdynamic', '-L/usr/lib', '-L/usr/lib/x86_64-linux-gnu', '-lboost_serialization', '-lboost_filesystem', '-lboost_system', '-lboost_unit_test_framework', '-lboost_program_options', '-lboost_graph', '-lboost_thread', '-lboost_regex', '-ltbb']
 -> task in 'test_cvt_map_elites' failed with exit status 1: 
	{task 140195946492200: cxxprogram test_cvt_map_elites.cpp.1.o -> test_cvt_map_elites}
['/usr/bin/g++', 'modules/cvt_map_elites/test_cvt_map_elites.cpp.1.o', '-o/home/david/argos-sferes/sferes2/build/modules/cvt_map_elites/test_cvt_map_elites', '-Wl,-Bstatic', '-Lsferes', '-lsferes2', '-Wl,-Bdynamic', '-L/usr/lib', '-L/usr/lib/x86_64-linux-gnu', '-lboost_serialization', '-lboost_filesystem', '-lboost_system', '-lboost_unit_test_framework', '-lboost_program_options', '-lboost_graph', '-lboost_thread', '-lboost_regex', '-ltbb']
`

**Comment by daneshtarapore** [comment]
Thanks for the centroid files. But I still can't replicate your error:

danesh@danesh-UOS214035:~/argos-sferes$ ./bin/obsavoid_evolcvt42D experiments/mutualinfo.argos --load experiments/gen_100 -o outputfile -n 14
[INFO] Not using threads
[INFO] Using random seed = 56
[INFO] Using simulation clock tick = 0.1
[INFO] Total experiment length in clock ticks = 1200
[INFO] Loaded library "./lib/libobsavoid_nn_controller.so"
[INFO] Loaded library "./lib/libobsavoid_evol_loopfunctionscvt42D.so"
[INFO] The physics engine "dyn2d" will perform 10 iterations per tick (dt = 0.01 sec)
[INFO] No visualization selected.

Loaded 1000 centroids.
sferes2 version: (const char*)"f731bfb04a48a475dfedbfdb180d77054498557f"
seed: 1558020733
loading experiments/gen_100
terminate called after throwing an instance of 'boost::archive::archive_exception'
  what():  unsupported version
Aborted (core dumped)

Am using Boost version: 1.65.1. Also, when trying to run an evolutionary experiment, I get this error:

danesh@danesh-UOS214035:~/argos-sferes$ ./bin/obsavoid_evolcvt42D experiments/mutualinfo.argos 
[INFO] Not using threads
[INFO] Using random seed = 56
[INFO] Using simulation clock tick = 0.1
[INFO] Total experiment length in clock ticks = 1200
[INFO] Loaded library "./lib/libobsavoid_nn_controller.so"
[INFO] Loaded library "./lib/libobsavoid_evol_loopfunctionscvt42D.so"
[INFO] The physics engine "dyn2d" will perform 10 iterations per tick (dt = 0.01 sec)
[INFO] No visualization selected.

Loaded 1000 centroids.
sferes2 version: (const char*)"f731bfb04a48a475dfedbfdb180d77054498557f"
seed: 1558015730
writing...archive_0
./bin/obsavoid_evolcvt42D_2019-05-16_08_46_51_22919/gen_0 written
writing...archive_1
./bin/obsavoid_evolcvt42D_2019-05-16_08_46_51_22919/gen_1 written
writing...archive_2
./bin/obsavoid_evolcvt42D_2019-05-16_08_46_51_22919/gen_2 written
writing...archive_3
./bin/obsavoid_evolcvt42D_2019-05-16_08_46_51_22919/gen_3 written
writing...archive_4
./bin/obsavoid_evolcvt42D_2019-05-16_08_46_51_22919/gen_4 written
terminate called after throwing an instance of 'std::runtime_error'
  what():  normalised MI should be in [0,1]
Aborted (core dumped)

**Comment by bossdm** [comment]
that is the same error as I used to get. it is because iridis uses 1.66. I solved it by upgrading to 1.70 though any version >=1.66 should work.

**Comment by bossdm** [comment]
the second one is strange: i am running iridis for more than a day and it is still going. I will have a closer look what could be happening.

**Comment by bossdm** [comment]
in meanwhile, can you try to replicate the error by taking one of the generation files you have, and testing an existing individual ?

**Comment by bossdm** [comment]
I am starting to think that maybe cvt_map_elites does not support loading NN individuals.

the only reference to load I could find is in cvt_binary_map.hpp, where they say it only works for EvoFloat and Sampled genotypes.

**Comment by bossdm** [comment]
I cannot replicate the error that you found:
`./bin/obsavoid_evolcvt42D_2019-05-16_08_46_51_22919/gen_4 written
terminate called after throwing an instance of 'std::runtime_error'
what(): normalised MI should be in [0,1]
Aborted (core dumped)`

On my laptop I am at generation 8 and no error, with the same seed. Iridis is running fine as well.

 I did strangely start to experience segmentation faults after I tried to install sferes in the argos-sferes folder. They disappeared after installing in a new folder the argos-sferes again according to the documentation.
